src/components/asthma/helpers/asthma-data.ts

```diff
@@ -0,0 +1,343 @@
+};
+
+const computeAirQualityStatusAndDescription = (aqi: number | undefined, defaultStatus: AsthmaDataStatus): [AsthmaDataStatus, string | undefined] => {
+    if (aqi) {
+        if (aqi <= 100) {
+            return ['in-range', undefined];
+        }
+        if (aqi <= 200) {
+            return ['out-of-range', 'Unhealthy'];
+        }
+        if (aqi <= 300) {
+            return ['out-of-range', 'Very Unhealthy'];
+        }
+        if (aqi <= 500) {
+            return ['out-of-range', 'Hazardous'];
+        }
+    }
+    return [defaultStatus, undefined];
+};
+
+const computeAirQuality = (type: AsthmaAirQualityType, dataPoints: DeviceDataPoint[], computeAqi: (dataPoints: DeviceDataPoint[]) => number, defaultStatus: AsthmaDataStatus): AsthmaAirQuality => {
+    let filteredDataPoints = dataPoints.filter(dp => dp.type === airQualityTypeTermCodes[type]);
+
+    let aqi: number | undefined;
+    if (filteredDataPoints.length > 0) {
+        aqi = computeAqi(filteredDataPoints);
+    }
+
+    let [status, description] = computeAirQualityStatusAndDescription(aqi, defaultStatus);
+
+    return {
+        type: type,
+        status: status,
+        value: aqi,
+        description: description
+    }
+};
+
+
+export interface AsthmaDataService {
+    loadParticipant(): Promise<AsthmaParticipant>;
+
+    loadDeviceInfo(): Promise<DeviceInfo>;
+
+    loadLogEntries(fromDate?: Date, toDate?: Date): Promise<AsthmaLogEntry[]>;
+
+    loadBiometricsForControlStatus(): Promise<AsthmaBiometric[]>;
+
+    loadBiometricsForDate(date: Date): Promise<AsthmaBiometric[]>;
+
+    loadAirQualitiesForControlStatus(homeAirQualityZipCode: string, workAirQualityZipCode: string): Promise<AsthmaAirQuality[]>;
+
+    loadAirQualitiesForDate(date: Date): Promise<AsthmaAirQuality[]>;
+
+    saveLogEntry(logEntry: AsthmaLogEntry): Promise<void>;
+
+    loadAndClearAlertTakeover(): Promise<string | undefined>;
+
+    loadSurveyAnswers(surveyNames: string[]): Promise<SurveyAnswer[]>;
+
+    loadAsthmaActionPlan(): Promise<AsthmaActionPlan | undefined>;
+}
+
+const service: AsthmaDataService = {
+    loadParticipant: async function (): Promise<AsthmaParticipant> {
+        let participantInfo = await MyDataHelps.getParticipantInfo();
+        return new AsthmaParticipant(participantInfo);
+    },
+    loadDeviceInfo(): Promise<DeviceInfo> {
+        return MyDataHelps.getDeviceInfo();
+    },
+    loadLogEntries: async function (fromDate?: Date, toDate?: Date): Promise<AsthmaLogEntry[]> {
+        let params: DeviceDataPointQuery = {
+            namespace: 'Project',
+            type: ['Asthma-LogEntry']
+        };
+        if (fromDate) {
+            params.observedAfter = fromDate.toISOString();
+        }
+        if (toDate) {
+            params.observedBefore = toDate.toISOString();
+        }
+        let result = await MyDataHelps.queryDeviceData(params);
+        return result.deviceDataPoints.map(dp => JSON.parse(dp.value) as AsthmaLogEntry);
```

Comment:
There are a lot of .finds against this data. An object/map keyed by the date used in finds would be more direct

Reply:
I'm not sure it would be more efficient to always have a map.

There are quite a few places where `.find` is used against this returned list, but not usually in the course of the single load/render cycle.  In other words, when this data is loaded for any particular component, usually there is only a single search of the list, which can short circuit on the first element that matches.  Converting the array to a map would always require looping through the entire list of entries, while building up a larger in-memory data structure.  The single find per load/render only has to loop the entire array if the entry it is searching for happens to be the last one, and no additional memory is used while doing so.

There is one spot in the calendar where more than one of the entries is to be rendered, and it would require multiple iterations on the list to find each one each time.  In that situation, a map of identifier-to-entry is being built and used.

Reply:
Can you please link me to the spot you are referring to?